MedHelm: Implement medcalc bench scenario, metrics and specs

sashimono-san opened the pull request:
This PR has the adds support to **MedCalc-Bench** benchmarking.

As much as possible, I maintained the implementation as done originally. For example, data is downloaded directly from [huggingFace](https://huggingface.co/datasets/ncbi/MedCalc-Bench-v1.0), and that the evaluation metric is _basically_ a copy of the [original benchmark repo](https://github.com/ncbi-nlp/MedCalc-Bench). But in some cases, changes were necessary, as discussed in the following paragraphs.

The [original benchmark repo](https://github.com/ncbi-nlp/MedCalc-Bench)) expects the model to answer in JSON format. But considering we do not want to necessarily evaluate models' ability to output JSON, prompts were adapted so the output is given in natural language. It would be great if someone can review the prompts, make sure they align with other Helm scenarios.

Originally, one-shot examples come from a pre-defined JSON, and a specific example is used for each question type (defined by the field `Calculator Id` in the dataset. I couldn't figure how to pass information, at runtime, from the `Scenario` to the `RunSpec`. Namely I did not find a way to extract from the `Instance` being used the `Calculator Id` and use it when building the inference prompt. **Is this supported by the current implementation of Helm?**

The original implementation has a special truncation logic for the one-shot method. [They truncates the Patient note and the step by step explanation of the output depending on the model](https://github.com/ncbi-nlp/MedCalc-Bench/blob/048ba77dbe332e9190935e4a30965bff444b940e/evaluation/run.py#L201). This is currently not implemented, but already during local tests (with gpt2) I had issues with input length. Any suggestions on how to solve this?

sashimono-san commented:
Still a WIP. Opened the PR by mistake, apologies.

yifanmai commented:
Thanks for your pull request! This is taking me longer than usual to review due to the size, but I will get back to you by this week.

yifanmai commented on src/helm/benchmark/scenarios/medcalc_bench_scenario.py:
```diff
@@ -0,0 +1,96 @@
+                            "lower_limit": example["Lower Limit"],
+                            "upper_limit": example["Upper Limit"],
+                            "calculator_id": example["Calculator ID"],
+                            "ground_truth": example["Ground Truth Answer"],
```
Delete `ground_truth` and use the answer in the reference instead.

yifanmai commented on src/helm/benchmark/scenarios/medcalc_bench_scenario.py:
```diff
@@ -0,0 +1,96 @@
+                        split=helm_split_name,
+                        extra_data={
+                            "id": example["Row Number"],
+                            "relevant_entities": example["Relevant Entities"],
```
Delete `relevant_entities` since it doens't seem to be used.

yifanmai commented on src/helm/benchmark/scenarios/medcalc_bench_scenario.py:
```diff
@@ -0,0 +1,96 @@
+                        ],
+                        split=helm_split_name,
+                        extra_data={
+                            "id": example["Row Number"],
```
Delete `id` - instead just set `id="id" + example["Row Number"]` on the `Instance` itself.

yifanmai commented on src/helm/benchmark/scenarios/medcalc_bench_scenario.py:
```diff
@@ -0,0 +1,96 @@
+
+    HUGGING_FACE_DATASET_PATH: str = "ncbi/MedCalc-Bench-v1.0"
+
+    # TODO: Add a base url
+    DATASET_DOWNLOAD_BASE_URL: str = ""
```
Looks unused; remove?

yifanmai commented on src/helm/benchmark/run_specs/medcalc_bench_specs.py:
```diff
@@ -0,0 +1,132 @@
+                "set their adjusted body weight to the minimum of the ideal body and actual weight. If the "
+                "patient is underweight, please set their adjusted body weight to their actual body weight."
+            ),
+            calculator_id="2",
```
There's a couple of ways to get this to work:

1. Write a custom adapter that looks up the calculator id and prepends the correct example.
2. Move this logic into the Scenario i.e. add an `one_shot` argument to the scenario, and when it is `True`, look up the corresponding example in the JSON file and prepend it to `text` inside `Input` inside `Instance`.

I think the second method is easier and more straightforward, so I'd recommend doing that.

sashimono-san replied:
I went with 2, it's implemented in the latest commit (along with the other change requests)

yifanmai commented on src/helm/benchmark/run_specs/medcalc_bench_specs.py:
```diff
@@ -0,0 +1,132 @@
+    )
+
+
+def _get_zero_shot_cot_instructions() -> str:
```
What's the rationale for using CoT here compared to the original scenario? Is it because we expect models to use CoT for calculations in realistic settings?

sashimono-san replied:
For us there's no practical difference as we only implement the CoT prompt anyway. I kept the naming from the [original benchmark repo](https://github.com/ncbi-nlp/MedCalc-Bench) so we can track future changes easily in the future.

From their paper, they compare performance of three approaches: direct prompting, CoT, and code generation.

I didn't implement the code generation testing because it seemed to add another focus to the evaluation (not only medical knowledge, but also programming knowledge). And I opted for CoT instead of direct prompting because it is rather well stablished CoT helps models make correct calculations, even though it deviates from the original motto of testing models and not prompt techniques.

yifanmai commented on src/helm/benchmark/run_specs/medcalc_bench_specs.py:
```diff
@@ -0,0 +1,132 @@
+    The original expect that for each sample, we collect the calculator ID and the question for building the one-shot instructions.
+    """
+    examples: Dict = {}
+    with open(ONE_SHOT_EXAMPLES_URL, "r") as f:
```
If you move this inside the scenario as I suggested above, you should also replace this raw `open()` with `ensure_file_downloaded()` instead.

yifanmai commented:
Regarding outputting JSON, I'm okay with changing this to not output JSON.

Regarding truncation, ideally I would prefer to avoid this issue altogether by using models with longer context lengths (GPT-2 only has 1k tokens). If you're using recent models only, almost all recent models have context lengths of at least 8k, so truncation might not be needed. But if you need truncation, then you would need to implement a custom adapter.

sashimono-san commented:
> u're using recent models only, almost all recent models have context lengths of at least 8k, so truncation might not be needed. But if you need truncation, then you would need to implement a custom adapter

Truncation is being handled separately by the medhelm, so I'll leave this out of this PR

sashimono-san commented:
@yifanmai I implemented and tested the changes you requested. Thank you for clarifying how to approach the one shot prompt generation.

As you might have noticed, I have changed the PR target branch to `med-helm`, talking to the helm team, they tell me `med-helm` will be reviewed later in bulk. And thus they instructed us to commit directly into `med-helm`. Since I had already opened this PR, I implemented the changes here. With that, feel free to review the new changes or merge it directly into `med-helm` for later review.
